runtime/random_numbers.cu

```diff
@@ -51,6 +51,42 @@ __device__ uint4 philox(
   return output;
 }
 
+template <unsigned int significand_bits>
```

Comment:
Why in `uniformf` we are using 32, and `uniform` we use `53`? Shouldn't we be using `24` and `53`?

Reply:
That's a good question.

Note that `uniform` is not exactly equivalent to `uniform_for_casting<53>(z)`. `uniform_for_casting` uses the full range of available integers instead of bit masking and only changes the scale so that the largest sampled value is not rounded to 1. `uniform` just sets the scale factor to 2^(-53). These scale factors should be similar but not quite equal. Note that using the full set of bits is important wrt quantization of samples near 0 (you can see that in the current approach because our lowest samples for bfloat are not 0.0).

For `uniformf`, we use a scale factor of 2^(-32) which means the largest sample we can return is $(2^{32} -1)/2^32 + 2^{-33}=1-3.49e-10$. However `nextafter(1, 0)` for `torch.float32` is `1-5.96e-8`. Sure enough:
```
[nav] In [31]: (torch.tensor(2**32 - 1, dtype=torch.float32) * torch.tensor(2.3283064e-10, dtype=torch.float32)).item()
Out[31]: 1.0
```
The reason we don't output 1.0 is this line:
```c++
return result == 1 ? 0.0f : result;
```
The intent is to guarantee the range is correct, but that should instead by done via rejection sampling (if at all). This kind of reassignment actually upweights 0.0 so that it's slightly more probably than it should be. Instead, I think we actually should use `uniform_for_casting<24>` for float and remove the ternary op.

Reply:
> This kind of reassignment actually upweights 0.0

Isn't the `+ scale / 2` here to prevent this issue? Because of this `+ scale / 2`, we never generate 0. So reassigning 1 as 0 will make sure 0 has the same probability as other values.

Reply:
I thought `scale/2` was to preserve the expected value. Reassigning 1 as 0 does not guarantee that we'll sample any other values between 0 and `scale / 2` for example. Is there value in generating 0.0? Otherwise we can scale properly so that we naturally sample right up to `nextafter(1, 0)`; it just means we won't sample some very small values. Note that we'll never sample lots of values anyway unless we use a bigger int input. For example if we use 64-bit input instead of 32, our scale factor for `float` will be 2.7105052696550477e-20 instead of 1.1641531488804091e-10. But we'll never feasibly sample `nextafter(0, 1, float) = 1.4013e-45` for example since we'd need something like 150 bits of samples.

Reply:
In light of offline discussion, I'm going to alter this PR to simply move the `result == 1 ? 0 : result` to _after_ the downcast. This should better match pytorch+curand.